roles/openshift_master_certificates/tasks/csr.yml

```diff
@@ -0,0 +1,238 @@
+  delegate_to: "{{ item }}"
+  run_once: true
+
+- name: Create the CSR for openshift-master
+  shell: >
+    /bin/openssl req -multivalue-rdn -nodes -newkey rsa:2048 -keyout {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-master.key
+    -out {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-master.csr
+    -config {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-master-openssl.cnf
+    -reqexts req_ext
+  with_items: "{{ hostvars
+                  | lib_utils_oo_select_keys(groups['oo_masters_to_config'])
+                  | lib_utils_oo_collect(attribute='inventory_hostname') }}"
+  delegate_to: "{{ item }}"
+  run_once: true
+
+- name: Create the CSR for openshift-aggregator
+  shell: >
+    /bin/openssl req -multivalue-rdn -nodes -newkey rsa:2048 -keyout {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-aggregator.key
+    -out {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-aggregator.csr
+    -config {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/openshift-aggregator-openssl.cnf
+    -reqexts req_ext
+  with_items: "{{ hostvars
+                  | lib_utils_oo_select_keys(groups['oo_masters_to_config'])
+                  | lib_utils_oo_collect(attribute='inventory_hostname') }}"
+  when: item == groups.oo_first_master.0
+  delegate_to: "{{ item }}"
+  run_once: true
+
+- name: Create the CSR for master.etcd-client
+  shell: >
+    /bin/openssl req -multivalue-rdn -nodes -newkey rsa:2048 -keyout {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/master.etcd-client.key
+    -out {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/master.etcd-client.csr
+    -config {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/etcd-client-openssl.cnf
+    -reqexts req_ext
+  with_items: "{{ hostvars
+                  | lib_utils_oo_select_keys(groups['oo_masters_to_config'])
+                  | lib_utils_oo_collect(attribute='inventory_hostname') }}"
+  delegate_to: "{{ item }}"
+  run_once: true
+
+- name: Create the CSR for etcd.server
+  shell: >
+    /bin/openssl req -multivalue-rdn -nodes -newkey rsa:2048 -keyout {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/etcd.server.key
+    -out {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/etcd.server.csr
+    -config {{ openshift_csr_dir }}/master-{{ hostvars[item].openshift.common.hostname }}/etcd.server-openssl.cnf
+    -reqexts req_ext
+  with_items: "{{ hostvars
+                  | lib_utils_oo_select_keys(groups['oo_masters_to_config'])
+                  | lib_utils_oo_collect(attribute='inventory_hostname') }}"
+  when: item == groups.oo_first_master.0
+  delegate_to: "{{ item }}"
+  run_once: true
```

Comment:
Not specific to this point, but after a while I realized we have a bunch of delegate_do + run_once options.
I have a feeling you may want a conditional run, based on the presence of the output file or not, rather than run_once everywhere ? Wouldn't that be a more robust pradigm ? As it would allow to rerun the playbook if it failed midway and avoid recreating CSRs that have already been created ?

Reply:
Let's leave in temp w/ mktemp.  We do this with other roles.

Reply:
fine